Fix: Blank screen after creating an account

kidroca opened the pull request:
<If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit.>
@marcaaron

### Details
Address the problem that no navigator is available during application initialization
We don't need to keep the entire AuthScreens navigator unmounted but just the Report screen
This way the URL still resolves to the correct report after the data is available

### Fixed Issues
Fixes #2497

### Tests
<!--- 
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->

1. Start from logged out state and proceed to create a new account
2. While logged out, type and login with brand new email address so it can trigger sign up flow
3. Get validation link from e-mail. On dev I'm copying the email link and appending it to localhost - e.g. http://localhost:8080/setpassword/10024706/EHOHMPGWM
4. Type new password and click "set password"
5. After a brief loading time you should see an initial chat list in LHN containing only the Concierge chat
    - the URL in the address would be `/r` until the initial report number is resolved
    - Depending on internet speed this can be too quick to notice
    - In the video samples below it reads `/loading` but we decided to change it
6. The Concierge chat should be loaded as the active conversation

Please check the steps listed in the PR here #2346 work as well and there's no regressions

### QA Steps
Same as above

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

I'm intentionally setting a slow internet connection to test that the URL (chat) would resolve correctly even when the data comes after the App's initial render

##### Creating account over a slow internet connection
https://user-images.githubusercontent.com/12156624/116293730-db8cab80-a79f-11eb-9ace-9f8ff0d4047f.mp4

##### Login in over a slow internet connection

https://user-images.githubusercontent.com/12156624/116294035-2c9c9f80-a7a0-11eb-84d2-de53b20003f3.mp4

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

##### Logging in on mobile web
https://user-images.githubusercontent.com/12156624/116302433-d7658b80-a7a9-11eb-814d-947bdeec626b.mov

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

##### Logging in on the desktop app
https://user-images.githubusercontent.com/12156624/116306650-0251de80-a7ae-11eb-8b93-e269f128f7df.mov

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

A simulated login using the magic link after an account is created

##### Simulate a magic link to set a password for a new account

https://user-images.githubusercontent.com/12156624/116302029-61f9bb00-a7a9-11eb-9a06-58b4d59071d9.mp4

#### Android

Testing on Android to verify everything still works as before

https://user-images.githubusercontent.com/12156624/116294183-58b82080-a7a0-11eb-960c-7802634bbe6f.mp4

kidroca committed:
feat: Navigation navigateLater and enableNavigation

Wait for the NavigationContainer and all its children finish mounting for the first time
before allowing/performing navigation

If there were navigation attempts - the last navigation attempt would be
automatically applied when the navigation is enabled

kidroca committed:
refactor: AuthScreens empty string is a valid initial value for initialReportID

Sometimes `currentReportID` might resolve as empty '' and prevent the
RootStack navigator from mounting

`const currentReportID = firstReportID ? String(firstReportID) : '';`

https://github.com/Expensify/Expensify.cash/blob/0da357c30bc1ace6d354a81aabe075f87c12144c/src/libs/actions/Report.js#L688

kidroca committed:
fix: Report.fetchAllReports - fetchChatReportsByIDs would be called when there are no reports

When `response.chatList` is an empty array (new users) - `reportIDs = String(response.chatList).split(',');`

Results in `reportIDs` being equal to `['']` and the former control flow `if (reportIDs.length)`
evaluated to `true`

Not casting the chat list IDs to string is not longer an issue because all the following usages
are working with strings as well as with numbers

kidroca committed:
fix: Report.fetchAllReports - new report created by `fetchOrCreateChatReport` is not used in `reportIDs`

kidroca committed:
fix: Report.fetchAllReports - Navigation.navigate does not work when no router is mounted

kidroca commented on src/libs/Navigation/Navigation.js:
```diff
@@ -91,6 +92,27 @@ function dismissModal(shouldOpenDrawer = false) {
+    // eslint-disable-next-line no-use-before-define
+    exports.navigate = navigate;
+}
+
```
This didn't resolve the original issue, and perhaps would not be a useful addition

Here's what happens following the Test steps in the PR

1. Start logged out - The Navigation is immediately enabled because we have an active navigator - `PublicScreens`
2. While logged out, type and login with brand new email address so it can trigger sign up flow
3. Get validation link from e-mail. On dev I'm copying the email link and appending it to localhost - e.g. http://localhost:8080/setpassword/10024706/EHOHMPGWM
4. Type new password and click "set password" - at this point `PublicScreens` unmount, but `AuthScreens` are not yet mounted. An error happens as a call to `navigate` is made during this time

marcaaron replied:
Sounds like we'd need to hook into some other `onReady` situation like the sidebar mounting or something to determine "readiness". But if we can avoid this complexity then let's do that instead :)

kidroca commented on src/libs/actions/Report.js:
```diff
@@ -659,9 +659,15 @@ function fetchOrCreateChatReport(participants, shouldNavigate = true) {
             if (shouldNavigate) {
-                // Redirect the logged in person to the new report
-                Navigation.navigate(ROUTES.getReportRoute(data.reportID));
+                try {
+                    // Redirect the logged in person to the new report
+                    Navigation.navigate(ROUTES.getReportRoute(data.reportID));
+                } catch (e) {
+                    // When switching between the Auth and Public routers there's a brief time of no navigator mounted
+                    console.error(e);
+                }
             }
```
This is executed after the Navigation was initialized (when `PublicScreens` mounted) but before the `AuthScreens` are mounted and throws an error. When the error is not handled the promise chain is broken and the original caller never resolves which in turn prevents rendering the initial state of the app

kidroca replied:
This would be reverted as well if we use a different approach

kidroca commented on src/libs/actions/Report.js:
```diff
@@ -727,15 +733,16 @@ function fetchAllReports(
                 return;
             }
 
-            // The string cast here is necessary as Get rvl='chatList' may return an int
-            reportIDs = String(response.chatList).split(',');
```
This produced a bug when `response.chatList` is empty `[]`. It would assign `['']` to report IDs, and the check ` if (reportIDs.length)` would be positive even though there are no reports

You can try this to see the outcome is `['']`

```js
console.log(String([]).split(',')) // prints ['']
```

---

Casting to string is no longer necessary as following usages of `reportIDs` work with both strings and numbers.
But if you need to have string IDs you can instead use map

```js
reportIDs = response.chatList.map(n => String(n)) // or n.toString()
```

marcaaron replied:
To clarify, the string cast is necessary because the API can return a `Number` (could be fixed eventually, but we are working around it for now).

kidroca replied:
Yes but the only usages have no problems with numbers:

`fetchChatReportsByIDs(reportIDs);` would convert everything to string anyway since internally it makes this `chatList.join(',')`

`fetchActions(reportID);` seems to work OK with number or string

And merging to Onyx would again use `String`

```js
const currentReportID = firstReportID ? String(firstReportID) : '';
```

marcaaron replied:
I'm still a little confused sorry. When we get the `response.chatList` from `API.Get` it can be a number or a string and must be turning into an `Array`. `fetchChatReportsByIDs()` is looking for an `Array` and it calls `.join()` on it. So it's not clear how this code is working.

kidroca replied:
For me `response.chatList` is always an Array containing Integers like so:
![image](https://user-images.githubusercontent.com/12156624/116244829-f7c52400-a770-11eb-8f85-ed280a2f8232.png)

Or an empty array like `[]` for new users

Are you saying there can be cases:
- a single number like `71402263` is returned
- a comma separated string is returned like `"71402263,71403817,71403849,71734617,71955477,72156283,72207410"`

In that case, yeah a special handling would be needed

marcaaron replied:
I may have misspoke, but I am remembering that it is "weird". 😂 

I think specifically it can be either an `Array` (like you are seeing) or a single number. Hence, why we were casting the array into a string before splitting it back into an array.

kidroca replied:
Ok, in that case I'll revert the old logic, but add filtering on the resulting array so that the bug with the false positive is addressed. 

This is recent though as before, creating account would just fail due to the navigation issue, and then if you refresh everything is fine, while now if this is not addressed you will never enter the case where a Concierge conversation is created and you're stuck with no chats

kidroca reviewed (commented):
I think it might be better to remove the conditional rendering in `AuthScreens`:

https://github.com/kidroca/Expensify.cash/blob/c4d25f5ed8c92131d48adcfa9affeee9baf75c86/src/libs/Navigation/AppNavigator/AuthScreens.js#L170-L174

And let the `AuthScreens` mount instantly so that stack navigators can swap instantly and a navigator is always available

This would also remove the code inside `shouldComponentUpdate`:

https://github.com/kidroca/Expensify.cash/blob/c4d25f5ed8c92131d48adcfa9affeee9baf75c86/src/libs/Navigation/AppNavigator/AuthScreens.js#L151-L158

When the app initializes from a stored state it would still load the persisted report
If no report is available in route params by the time the `ReportScreen` is mounted it can request navigation to transition to the first available report in the chat list

This way most of the changes in this PR, but the bug fixes, will be removed

kidroca commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -169,7 +169,7 @@ class AuthScreens extends React.Component {
 
     render() {
         // Wait to resolve initial Home route params.
-        if (!this.initialReportID) {
+        if (!_.isString(this.initialReportID)) {
             return null;
         }
```
What I'm trying to say here: https://github.com/Expensify/Expensify.cash/pull/2575#pullrequestreview-644939666 is most of the issues would come when we return `null` here and there's no mounted navigator

The code would become simpler with `initialReportID` defaulting to empty string, and when no stored report is available during init we still mount the navigator, when the ReportScreen mounts it can make a call to `Navigation.navigate(/* any/first available report */)` if there's nothing available **and the screen is focused** - e.g. it would not mess navigation if we're deeplinked to some other screen

marcaaron replied:
That makes sense. But to back up for a second, wasn't the point of waiting so that we could pass the `initialReportID` to `initialParams` of the screen?

kidroca replied:
Yes, but this waiting is causing issue like the one we're trying to solve. 

By removing it we'll won't have to come up with solutions like `try/catch` on certain `navigate` calls or custom nav handling like `navigateLater` 

Initial Prams was a good solution to fix the url in the address bar, but it's not the only one
I'll make sure to link the previous PR as well so that there's no regression on that matter

kidroca committed:
refactor: Move initialReportID from AuthScreens to MainDrawerNavigator

kidroca committed:
refactor: Mount the Report Screen only when initial data is resolved

kidroca committed:
refactor: Reuse `updateCurrentlyViewedReportID`

kidroca committed:
Revert "feat: Navigation navigateLater and enableNavigation"

This reverts commit 529f0fb6

These changes didn't prove as useful as intended

kidroca committed:
Revert "fix: Report.fetchAllReports - Navigation.navigate does not work when no router is mounted"

This reverts commit c4d25f5e

This fix is no longer needed

kidroca commented on src/libs/Navigation/AppNavigator/MainDrawerNavigator.js:
```diff
@@ -28,22 +41,37 @@ const MainDrawerNavigator = props => (
-            }}
-        />
+        {
+            _.isString(props.initialReportID)
+                ? (
+                    <Drawer.Screen
+                        name="Report"
+                        component={ReportScreen}
+                        initialParams={{reportID: props.initialReportID}}
+                    />
+                )
+                : (
+                    <Drawer.Screen name="loading">
+                        {() => <FullScreenLoadingIndicator visible />}
+                    </Drawer.Screen>
+                )
```
Moving the conditional mounting here allows to always have a navigator mounted.

The Report screen will mount once `initialReportID` is available - either instantly the usual case, or after a brief time - login / create account.

In the meantime a placeholder screen is mounted as you can't have a Navigator without screens. It doesn't need to render anything, but I've added the loader. As a name I've used "loading" this would appear as a URL until the Report Screen is mounted, this can be renamed or setup so that it doesn't appear in the URL. You can see this in the video below

I've used the `isString` check in the rare event that no `initialReportId` is not resolved - there's a case where `''` would be set. This is an error case because, as you know initialization logic would always set a `initialReportId` (or navigate to the Concierge conversation). In that case the view would just show the empty report page and you can recover by selecting a different conversation or refreshing the page.

This can perhaps be further improved a bit if state is used. E.g. save state the first time that `props.initialReportID` becomes available and keep the Report Screen mounted ever since that - in case you think that something could delete/remove the `props.initialReportID` e.g. Onyx eviction

##### Video simulating how initialReportId is resolved from a network call
https://user-images.githubusercontent.com/12156624/116292439-610f5c00-a79e-11eb-8283-5bf9ef3c4784.mp4
- You can also see the effect of deleting currently viewed report during runtime

1. Report saved in Onyx is deleted
2. App is reloaded from root `/`
3. The route (report ID) resolves from Network

thienlnam replied:
This logic makes sense 👍  
Though I think 

> can be renamed or setup so that it doesn't appear in the URL

we should do this as we don't want a url that has /loading

kidroca commented on src/libs/actions/Report.js:
```diff
@@ -950,7 +951,7 @@ function handleReportChanged(report) {
  */
 function updateCurrentlyViewedReportID(reportID) {
-    Onyx.merge(ONYXKEYS.CURRENTLY_VIEWED_REPORTID, String(reportID));
+    Onyx.merge(ONYXKEYS.CURRENTLY_VIEWED_REPORTID, String(reportID || ''));
```
I've updated this so it can be reused in `fetchAllReports` and to guard against `null` , `undefined` as `String(null)` would create a `"null"`  (similar situation for undefined and NaN)

marcaaron replied:
Hmm, this is the _one thing_ that I'm not so sure about. If we are calling this with a `reportID` that is `null` or `undefined` why try to set it at all? Couldn't we just leave whatever existing value we have? Seems not ideal to fail silently and set the `currentlyViewedReportID` to an empty string. Maybe it would be appropriate to log an alert of some kind when this happens instead?

We have some ways to do that e.g. 

https://github.com/Expensify/Expensify.cash/blob/f213be0a1be5a04599d2f17d8848b0a7824eb8be/src/libs/actions/Report.js#L717-L720

kidroca replied:
> Hmm, this is the one thing that I'm not so sure about. If we are calling this with a reportID that is null or undefined why try to set it at all? 

As you can see the original logic would save anything without any consideration, I've just updated it so that if called with null or undefined they are converted

I'll just revert this, this will also address the used-before-defined problem

marcaaron replied:
Yes, I see. Only mentioning this since logging an alert would perhaps be useful. But no need to add it in this PR.

kidroca commented on src/libs/actions/Report.js:
```diff
@@ -762,8 +763,8 @@ function fetchAllReports(
-            const currentReportID = firstReportID ? String(firstReportID) : '';
-            Onyx.merge(ONYXKEYS.CURRENTLY_VIEWED_REPORTID, currentReportID);
+            // eslint-disable-next-line no-use-before-define
+            updateCurrentlyViewedReportID(firstReportID);
```
Maybe I should move `updateCurrentlyViewedReportID` to be declared before this usage so the eslint override can be removed. 

About that eslint rule it can be tweaked. There's nothing wrong using a `function` before it's defined and the rule allows for such configuration. The problem is when you try to use a `class` or a variable before they are defined.
This style rule is forcing specifics to be defined at the top, which usually degrades readability.

marcaaron replied:
>About that eslint rule it can be tweaked.

It has been brought up before. We generally agree with the reasoning laid out here 

https://eslint.org/docs/rules/no-use-before-define

kidroca commented:
Updated

marcaaron commented:
Thanks, going to let @thienlnam take a first pass at reviewing + testing this then I will jump in.

thienlnam reviewed (commented):
Changes look good, just a small comment and then going to give it a test

kidroca commented:
I've added sample videos for all platforms:

I don't know how you test deep links on iOS / Android. This is what I do for iOS for example:

https://user-images.githubusercontent.com/12156624/116302029-61f9bb00-a7a9-11eb-9a06-58b4d59071d9.mp4
- create a new account
- copy the link from the email
- use the link in a terminal to trigger a simulator run: 
    ```shell
    xcrun simctl openurl booted expensify-cash://setpassword/10075909/XGFOJRHBO
    ```
    - note: you have to be logged out first in order for this particular link to resolve

Maybe something about it can be included in the README

kidroca committed:
refactor: Bring back the cast logic to fetchAllReports

kidroca committed:
refactor: Move Onyx updating code to promise finally

In the event of an error a currently viewed report would never get set
This would result in a invalid initial state of the app
Moving the logic to execute in `finally` would cover both
success and error cases and allow the app to continue and recover

kidroca committed:
fix: Remove `loading` from the URL path in the address bar

kidroca committed:
update: Keep the Report Screen mounted since the first time it's resolved

The Report Screen is setup to mount when:
- routing information is already available
- initial params provide fallback when routing information is missing

This would keep the Report Screen mounted in case the `initialReportID` is
ever reset to `null` like in the event of clearing Onyx data

kidroca committed:
refactor: Replace inline Screen names with values from SCREENS

kidroca commented on src/libs/Navigation/AppNavigator/MainDrawerNavigator.js:
```diff
@@ -1,22 +1,62 @@
+
 const Drawer = createDrawerNavigator();
 
+/**
+ * Derives whether it's time to mount the ReportScreen from current component props
+ *
+ * @param {String} initialReportID
+ * @param {Object} route
+ * @returns {boolean}
+ */
+const shouldMountReportScreen = ({initialReportID, route}) => {
+    if (_.isString(initialReportID)) {
+        return true;
+    }
+
+    // After the ReportScreen is mounted it should stay mounted no matter initial report ID
+    return Boolean(route.state) && route.state.routeNames.includes(SCREENS.REPORT);
+};
+
```
This could be put in the component body like

```jsx
const MainDrawerNavigator = (props) => {
    const hasInitialParams = _.isString(props.initialReportID);
    const const isReportMounted = Boolean(route.state) && route.state.routeNames.includes(SCREENS.REPORT);

   return (
        <Drawer.Navigator>
           {
               hasInitialParams || isReportMounted 
                  ? <Drawer.Screen name={SCREENS.REPORT} />
                  : <Drawer.Screen name={SCREENS.LOADING} />
            }
    );
```

---

Underscore `get` function would have been useful here

```js
return _.get(route, ['state', 'routeNames'], []).includes(SCREENS.REPORT);
```

But it's not available in the current version

kidroca replied:
I've just noticed there's a warning regarding accessing state
> Accessing the 'state' property of the 'route' object is not supported. If you want to get the focused route name, use the 'getFocusedRouteNameFromRoute' helper instead: https://reactnavigation.org/docs/screen-options-resolution/#setting-parent-screen-options-based-on-child-navigators-state

I'll see if it's possible to change this so this is not using `route.state`

kidroca commented on src/libs/actions/Report.js:
```diff
@@ -754,16 +759,15 @@ function fetchAllReports(
+                // eslint-disable-next-line no-use-before-define
+                updateCurrentlyViewedReportID(firstReportID);
             }
-
-            const firstReportID = _.first(reportIDs);
-            const currentReportID = firstReportID ? String(firstReportID) : '';
-            Onyx.merge(ONYXKEYS.CURRENTLY_VIEWED_REPORTID, currentReportID);
```
Moved this logic to `finally` as it's important to set the `CURRENTLY_VIEWED_REPORTID` during initialization, even in the event of an error

kidroca commented:
Updated

kidroca committed:
refactor: convert MainDrawerNavigator to class to avoid `route.state` usage

Keep the Report Screen mounted since the first time it's resolved

The Report Screen is setup to mount when:
- routing information is already available
- initial params provide fallback when routing information is missing

This would keep the Report Screen mounted in case the `initialReportID` is
ever reset to `null` like in the event of clearing Onyx data

kidroca commented:
Updated

kidroca committed:
fix: The logic moved to finally was not correctly translated

marcaaron commented on src/libs/Navigation/AppNavigator/MainDrawerNavigator.js:
```diff
@@ -1,49 +1,104 @@
+class MainDrawerNavigator extends React.Component {
+    constructor(props) {
+        super(props);
+        this.shouldMountReportScreen = _.isString(props.initialReportID);
```
It might not be clear that an `initialReportID` should always either exist or be on its way soon. Could we document this somehow?

marcaaron commented on src/libs/Navigation/AppNavigator/MainDrawerNavigator.js:
```diff
@@ -1,49 +1,104 @@
+            }
+        }
+
+        // Re-render the component only for these changes
```
Out of curiosity was the re-rendering causing some issues? Probably the `ReportScreen` is expensive to re-render so we typically want to avoid doing it?

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -727,15 +727,20 @@ function fetchAllReports(
+            // The cast here is necessary as Get rvl='chatList' may return an int or Array
+            reportIDs = String(response.chatList)
+                .split(',')
+                .filter(_.identity);
```
I'm not too sure what this `filter(_.identity)` does here? Do we need it?

kidroca replied:
`.filter(_.identity)` removes empty/null values from the resulting array. 
I thought it's clearer than `.filter(report => report)`

This is related to the bug I've commented about in the other thread: https://github.com/Expensify/Expensify.cash/pull/2575/files#r621584917

When `response.chatList` is an empty array `[]` (After creating an account for example) `String(response.chatList)` would return `['']` - an Array containing one empty string. Then the check below `reportIDs.length > 0` would redirect flow to `fetchChatReportsByIDs` and would not actually fetch anything. The LHN stays empty and Concierge chat is never created (Unless something else creates such a chat at a different time?)

marcaaron replied:
Ok, I think it's more common for us to use `filter(report => report)` or use `_.filter()` on it's own to remove bad values, but this isn't necessarily incorrect.

<img width="283" alt="Screen Shot 2021-04-29 at 7 13 48 AM" src="https://user-images.githubusercontent.com/32969087/116591425-82656a80-a8ba-11eb-8ac9-dc2aa13f3be8.png">

tgolen replied:
I came across this too, and I think it should be changed as it's not intuitive code (it's overly clever). Why would there be empty/null values? In what case would that ever happen?

kidroca replied:
> This produced a bug when `response.chatList` is empty `[]`. It would assign `['']` to report IDs, and the check ` if (reportIDs.length)` would be positive even though there are no reports
> 
> You can try this to see the outcome is `['']`
> 
> ```js
> console.log(String([]).split(',')) // prints ['']
> ```

`response.chatList` is `[]` when you create a new account 
when you convert that to a string `String(response.chatList)` and then split it `.split(',')`
you get `['']` and then the Concierge chat is not created

tgolen replied:
OK, thanks for explaining that! I think I have some ideas to clean this up.

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -727,15 +727,20 @@ function fetchAllReports(
+            return fetchOrCreateChatReport([currentUserEmail, 'concierge@expensify.com'])
+                .then((createdReportID) => {
+                    reportIDs = [createdReportID];
+                });
```
another option could be to have `fetchOrCreateChatReport()` return a reportID in an array and just use the `reportIDs` passed to the next chain. But this works (I had this idea from another PR I'm working on).

kidroca replied:
This would be changing the other function for the convenience of the current usage.
I wouldn't expect to receive a list of one element from a function like `fetchOrCreateChatReport`

marcaaron reviewed (dismissed):
Nothing in this PR is a major blocker for me so I am leaning towards merging so we can unblock the deploy and do a follow up to address any comments. cc @roryabraham

kidroca commented on src/libs/actions/Report.js:
```diff
@@ -754,16 +759,17 @@ function fetchAllReports(
+        .finally(() => {
             // Update currentlyViewedReportID to be our first reportID from our report collection if we don't have
             // one already.
             if (!shouldRedirectToReport || lastViewedReportID) {
                 return;
             }
```
Maybe consider changing the `shouldRedirectToReport` parameter. As you can see from this check - it will not be respected if `lastViewedRportID` is alredy set. 
Initially I thought that calling `fetchAllReports(shouldRedirectToReport: true)` will redirect to the report no matter what
But now it seems like it only serves to prevent a redirect

Instead of adding flags to the function you can do something like this in the only 2 usages:

##### Report.js
```js
NetworkConnection.onReconnect(() => {
    fetchAllReports();
});
```

##### AuthScreens
```jsx
class AuthScreens extends React.Component {
  componentDidMount() {
    this.initReports();
  }

  initReports() {
    fetchAllReports()
      .then(reportIds => {
         if (!this.props.lastViewedReportID) {
            const firstReportID = _.first(reportIDs);
            updateCurrentlyViewedReportID(firstReportID);
         }
      })
  }
}
```

I guess the concierge creating logic can be extracted as well.

What I'm saying is `fetchAllReports` is not only responsible for fetching reports but at least 2 other things.
If it's intended to always create the Concierge chat when there are no chats it can be renamed to something like `initUserReports` so the purpose is clear. From the current documentation the purpose of the function is just fetching

Even if the function is reused on the 10 places a specific usage like in the AuthScreens example can be extracted instead of flag parameters

kidroca replied:
I've actually found a different way and addressed this

kidroca committed:
refactor: remove memo usage

kidroca committed:
Revert "refactor: convert MainDrawerNavigator to class to avoid `route.state` usage"

This reverts commit a2aa335b

kidroca committed:
refactor: base home navigator mounting on whether there are any reports ready

There's no point in mounting the actual navigator before it has anything to route

Added a fallback logic here so that even if `initialReportID` is never set
the component would still resolve and function correctly

kidroca committed:
Revert "refactor: Reuse `updateCurrentlyViewedReportID`"

This reverts commit f394b98b

kidroca committed:
refactor: remove the `shouldRedirectToReport` and lastReportID setting

This handling is no longer needed
Initial routing data in `MainDrawerNavigator` would resolve with a reportID
Then the ReportScreen will store it in Onyx when it is mounted:
didMount -> storeCurrentlyViewedReport

kidroca committed:
feat: findLastAccessedReport a function to derive the last viewed report

When `initialReportId` is not available we can fall back to the last
viewed report.
BTW `initialReportId` might not be needed at all since it would be
storing the last accessed/viewed report

kidroca committed:
refactor: move `findLastAccessedReport` to `reportUtils`

When `initialReportId` is not available we can fall back to the last
viewed report.
BTW `initialReportId` might not be needed at all since it would be
storing the last accessed/viewed report

kidroca committed:
refactor: remove `initialReportID` no longer needed

`initialReportID` is always the same as the ID returned by `getLastAccessedReport`

kidroca commented on src/libs/Navigation/AppNavigator/MainDrawerNavigator.js:
```diff
@@ -1,49 +1,87 @@
-            initialParams={{reportID: ''}}
-            options={{
+// Decorated to always returning the result of the first call - keeps Screen initialParams from changing
+const getInitialReport = _.once(getLastAccessedReport);
```
`_.once` is ideal for initialization scenarios where the value need to be computed just once and always returned since

marcaaron replied:
Ok, so we are now waiting for the `reports` to exist instead of a `currentlyViewedReportID`. Makes sense.

kidroca commented on src/libs/Navigation/AppNavigator/MainDrawerNavigator.js:
```diff
@@ -1,49 +1,87 @@
+        return <FullScreenLoadingIndicator visible />;
+    }
+
+    const initialReportID = getInitialReport(props.reports).reportID;
```
Once reports are loaded initial value is derived and no longer updates.
It doesn't need to update because the Screen would use `initialParams` when it mounts and no longer care if they change

The value returned by `getLastAccessedReport` would be the same as `ONYXKEYS.CURRENTLY_VIEWED_REPORTID`, but if there's no data in Onyx (like after signout -> login) the last seen report by the user will be resolved, this seems a bit better than just using the first report in the list.

This also simplifies other places:
- nothing no longer needs to make sure a `lastViewedReportID` is set so that Navigation can mount
- A chunk of code in `Report.fetchAllReports` was removed as it's no longer necessary

marcaaron replied:
Yep that all sounds great to me!

kidroca commented on src/libs/actions/Report.js:
```diff
@@ -754,16 +757,6 @@ function fetchAllReports(
             // We are waiting 8 seconds since this provides a good time window to allow the UI to finish loading before
             // bogging it down with more requests and operations.
             }, shouldDelayActionsFetch ? 8000 : 0);
-
-            // Update currentlyViewedReportID to be our first reportID from our report collection if we don't have
-            // one already.
-            if (!shouldRedirectToReport || lastViewedReportID) {
-                return;
-            }
-
-            const firstReportID = _.first(reportIDs);
-            const currentReportID = firstReportID ? String(firstReportID) : '';
-            Onyx.merge(ONYXKEYS.CURRENTLY_VIEWED_REPORTID, currentReportID);
```
This is no longer needed because:
1. It was only called like that in `AuthScreens`
2. Initial report ID would be derived in `MainDrawerNavigator`
3. `MainDrawerNavigator` would mount the `ReportScreen`
4. `ReportScreen` would set `currentReportID` during mount and updates 
    https://github.com/Expensify/Expensify.cash/blob/42b73072b87ff9e22e2ce126c9b825c4096f6c50/src/pages/home/ReportScreen.js#L32-L35

marcaaron replied:
Ok so we will stop redirecting to the report and instead just show the most recently accessed report. Seems like a nice clean solution. Redirecting from here always felt a bit weird tbh.

kidroca reviewed (commented):
With these changes `ONYXKEYS.CURRENTLY_VIEWED_REPORTID` is no longer needed for initialization purposes.

kidroca committed:
fix: When `fetchAllReports` creates a chat with concierge it don't have to navigate to it

This would be resolved by initial params
Otherwise it still work, but an error is printed out that navigation failed
The ReportScreen is not mounted as the change that mounts it happens
a few lines before the call to navigate in `fetchOrCreateChatReport`

thienlnam reviewed (approved):
Nothing glaring is pointing out to me, let's get this deploy out - sending last review to you @marcaaron

marcaaron commented on src/libs/Navigation/AppNavigator/MainDrawerNavigator.js:
```diff
@@ -1,49 +1,87 @@
-            // eslint-disable-next-line react/jsx-props-no-multi-spaces
-            initialParams={{reportID: ''}}
-            options={{
+// Decorated to always returning the result of the first call - keeps Screen initialParams from changing
```
This comment is a little hard to follow and I might change it to something like:

>We are decorating `getLastAccessedReport` so that it caches the first result once the reports load. This will ensure the `initialParams` are only set once for the `ReportScreen`.

Which also begs the question... what happens if the `initialParams` change? And why do we care? Maybe we can explain that as well here just so it is very obvious?

kidroca replied:
Nothing will happen if we pass different `initialParams` after the screen is mounted. And that's why we want to stop calculating them each time `MainDrawerNavigator` re-renders

It might be better to just remove this `_.once` optimization and address it if it ever becomes a problem

marcaaron replied:
Nice one, yep I think this is fine in any case but took me a second to get why so a small comment would have been helpful.

roryabraham commented:
@kidroca, I noticed that you had one approval, and then made substantial changes to this PR. I don't have all the context for why those changes were made, but I did just want to point out that this PR fixes a [deploy-blocking issue](https://github.com/Expensify/Expensify.cash/issues/2497) that has been preventing us from running a production deploy for more than a week now. Generally with deploy blockers we want to get the quickest fix that solves the problem at hand so we're no longer holding up the deploy, even if you then create a follow-up PR to provide a better or more robust solution during the next deploy. Just something to keep in mind for the future 🙂

marcaaron reviewed (approved):
LGTM, I only have some suggestions about how to improve the documentation around some of the logic in the drawer navigator but no need to block.

kidroca commented:
@roryabraham 
I saw Marc comment regarding merging this.
It was a good point to explicitly say stop making further changes.

Since it wasn't merged and no one said nothing. I looked at the comments left by Marc and came up with an even more brilliant solution and had the time to implement it.
I always consider how easy would something be to review and wouldn't have done this if the solution wasn't way better
The changes might seem significant but 50% of them are blank lines comments and indentation change

Besides I've left a good description which should have helped reviewing them quickly

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/Expensify.cash/actions/runs/796947545) to staging in version: 1.0.33-4🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

marcaaron commented:
Just a heads up there is an issue with this implementation and it's _possible_ to get into a bad state if you have some partial report data e.g. a report that has only partial data. Here's what I've observed:

![Screen Shot 2021-04-29 at 4 07 05 PM](https://user-images.githubusercontent.com/32969087/116639811-12300680-a905-11eb-9b0b-535cbd2e6fcd.png)

Which, as you can imagine, leads to some pretty broken stuff 😄 

That said, I was only able to reproduce this a few times. Still we should be careful to avoid this case and not present any report that has no `reportID`.

kidroca commented:
Nice catch @marcaaron 

I didn't know that there could be such a report in the collection

I see that this is handled with a `reportID` check here: 

https://github.com/Expensify/Expensify.cash/blob/2ae23e75b9c7b9166fb2a333bfcdd10646508448/src/libs/OptionsListUtils.js#L192-L195

I don't know why incomplete reports aren't stored in a different collection but perhaps we can extract this check to `reportUtils` for the time being 

```js
function isIncompleteReport(report) {
  if (!report || !report.reportID) return true;

  const participants = lodashGet(report, ['participants'], []);
  return _.isEmpty(logins);
}
```

Then update the code here:

https://github.com/Expensify/Expensify.cash/blob/2ae23e75b9c7b9166fb2a333bfcdd10646508448/src/libs/reportUtils.js#L30-L35

To account for that

```js
function getLastAccessedReport(reports) {
    return _.chain(reports)
        .filter(report => isIncompleteReport(report) == false)
        .toArray()
        .sortBy('lastVisitedTimestamp')
        .last()
        .value();
}
```

I can't find how a partial report is created to analyze this further, can you point me to when/where this happens?

marcaaron commented:
Yep, perfect plan! I don't quite understand how they are created. It could be a race condition or a failed network request or something else entirely.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/Expensify.cash/actions/runs/822089210) to production in version: 1.0.39-5🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
